allennlp/models/graph_parser.py

```diff
@@ -0,0 +1,341 @@
+        arc_tags = arc_tags * arc_indices
+        arc_nll = self._arc_loss(attended_arcs, arc_indices) * float_mask.unsqueeze(1) * float_mask.unsqueeze(2)
+        # We want the mask for the tags to only include the unmasked words
+        # and we only care about the loss with respect to the gold arcs.
+        tag_mask = float_mask.unsqueeze(1) * float_mask.unsqueeze(2) * arc_indices
+
+        batch_size, sequence_length, _, num_tags = arc_tag_logits.size()
+        original_shape = [batch_size, sequence_length, sequence_length]
+        reshaped_logits = arc_tag_logits.view(-1, num_tags)
+        reshaped_tags = arc_tags.view(-1)
+        tag_nll = self._tag_loss(reshaped_logits, reshaped_tags.long()).view(original_shape) * tag_mask
+
+        valid_positions = tag_mask.sum()
+
+        arc_nll = arc_nll.sum() / valid_positions.float()
+        tag_nll = tag_nll.sum() / valid_positions.float()
+        return arc_nll, tag_nll
+
+    @staticmethod
+    def _greedy_decode(attended_arcs: torch.Tensor,
+                       arc_tag_logits: torch.Tensor,
+                       mask: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Decodes the head and head tag predictions by decoding the unlabeled arcs
+        independently for each word and then again, predicting the head tags of
+        these greedily chosen arcs indpendently. Note that this method of decoding
+        is not guaranteed to produce trees (i.e. there maybe be multiple roots,
+        or cycles when children are attached to their parents).
+
+        Parameters
+        ----------
+        attended_arcs : ``torch.Tensor``, required.
+            A tensor of shape (batch_size, sequence_length, sequence_length) used to generate
+            a distribution over attachements of a given word to all other words.
+        arc_tag_logits : ``torch.Tensor``, required.
+            A tensor of shape (batch_size, sequence_length, sequence_length, num_tags) used to
+            generate a distribution over tags for each arc.
+
+        Returns
+        -------
+        arc_probs : ``torch.Tensor``
+            A tensor of shape (batch_size, sequence_length, sequence_length) representing the
+            probability of an arc being present for this edge.
+        arc_tag_probs : ``torch.Tensor``
+            A tensor of shape (batch_size, sequence_length, sequence_length, sequence_length)
+            representing the distribution over edge tags for a given edge.
+        """
+        # Mask the diagonal, because we don't self edges.
+        inf_diagonal_mask = torch.diag(attended_arcs.new(mask.size(1)).fill_(-numpy.inf))
+        attended_arcs = attended_arcs + inf_diagonal_mask
+        # shape (batch_size, sequence_length, sequence_length, num_tags)
+        arc_tag_logits = arc_tag_logits + inf_diagonal_mask.unsqueeze(0).unsqueeze(-1)
+        # Mask padded tokens, because we only want to consider actual word -> word edges.
+        if mask is not None:
```

Comment:
You've already called `mask.size(1)` above, so you'll have crashed `if mask is None`.

Reply:
Thanks, fixed. The mask is actually required anyway.